les, les/flowcontrol: improved request serving and flow control

zsfelfoldi opened the pull request:
This PR
- replaces the existing request cost estimation method with a benchmark which gives much more consistent results
  - this also means the server can now estimate its own serving capacity properly and limit the number of accepted clients if necessary. Until now the allowed number of light peers was just a guess which probably contributed a lot to the fluctuating quality of available service.
- reimplements flowcontrol.ClientManager in a cleaner and more efficient way, with added capabilities:
  - better bandwidth control which allows using the flow control parameters for client prioritization
  - allowing target utilization over 100 percent at full load (parallel request processing allows this but the old client manager logic could not handle it)
  - reducing total serving bandwidth during block processing
- implements parallel LES request serving even for a single peer (a requirement for the new client manager logic)
- implements a simple private API for LES servers allowing server operators to assign priority bandwidth to certain clients and change prioritized status even while the client is connected
- adds a unit test for the new client manager
- adds an end-to-end test using the network simulator that tests bandwidth control functions through the new API

Note: there is a plan to implement in-protocol payment using the SWAP payment channel. This PR provides most of the functionality needed to implement such a feature. While the actual payment protocol is being developed the API can already be used and tested with other payment schemes (like a simple monthly subscription using on-chain payments).

Note 2: this PR is based on https://github.com/ethereum/go-ethereum/pull/17948

See also: https://gist.github.com/zsfelfoldi/93792da3c56dffdb594ff91aebd74262

shazow commented on les/api.go:
```diff
@@ -24,121 +24,121 @@ import (
 
 // PublicLesServerAPI  provides an API to access the les server.
 // It offers only methods that operate on public data that is freely available to anyone.
 type PrivateLesServerAPI struct {
```
Should this be `PrivateLESServerAPI` (acronyms capitalized)?

shazow commented on les/api_test.go:
```diff
@@ -51,34 +51,34 @@ request performance test. When testServerDataDir is empty, the test is skipped.
-// testBandwidthAPI runs an end-to-end simulation test connecting one server with
-// a given number of clients. It sets different priority bandwidths to all clients
+// testCapacityAPI runs an end-to-end simulation test connecting one server with
+// a given number of clients. It sets different priority capacitys to all clients
```
s/capacitys/capacities/ (also below)

zsfelfoldi replied:
done

shazow commented on les/api.go:
```diff
@@ -24,121 +24,121 @@ import (
-// If the assigned bandwidth is bigger than zero then connection is always
-// guaranteed. The sum of bandwidth assigned to priority clients can not exceed
-// the total available bandwidth.
+// SetClientCapacity sets the priority capacity assigned to a given client.
```
This explanation is a bit confusing for me. What does it mean to assign the capacity of a client? Sounds almost like we're setting the _priority_ of the client?

zsfelfoldi replied:
Priority has a slightly different meaning in my terminology. Being a priority client means you will never be disconnected and you can even connect if a non-priority client has to be kicked out to make room for you. Priorities can be compared but they are not scalar values. Capacity, on the other hand, is a scalar value that is proportional to the amount of requests served per time unit. It makes sense to buy priority even if you buy the same capacity as a free client gets (but you get it guaranteed for a long time).
Btw flow and capacity are established terms in graph theory and I think they can be applied here.
https://en.wikipedia.org/wiki/Flow_network

fjl commented on les/flowcontrol/manager.go:
```diff
@@ -18,207 +18,269 @@
-	defer self.lock.Unlock()
+// accepted deduces the upper estimate for request cost from the buffer and returns a priority
+// value based on current buffer status which is used by the serving queue.
+func (cm *ClientManager) accepted(node *ClientNode, maxCost uint64, now mclock.AbsTime) (priority int64) {
```
Documentation should say:

```text
accepted is called when a request with given maximum cost is accepted.
It returns a priority indicator for the request which is used to determine placement
in the serving queue. Priority is simple: older requests have higher priority. If the
client is almost out of buffer, request priority is reduced.
```

fjl commented on les/flowcontrol/manager.go:
Question for next review round: why is `sumRecharge` only updated conditionally? If the loop hasn't ended, `rcqNode` is done recharging. It seems dangerous to not remove it from `sumRecharge` in all cases.

zsfelfoldi replied:
Because it might have finished recharging earlier already. The rule here is that sumRecharge is always the sum of MinRecharge parameters of the nodes where corrBufValue < BufLimit. If that is not true here then a previous update has already subtracted MinRecharge from sumRecharge.

fjl commented:
About the benchmark: we agreed it is better to not run the benchmark on server startup. Instead, a debug API call should be added to run the benchmark. The request cost table should be hard-coded because it is an approximation of relative performance of different request types.

fjl commented on core/blockchain.go:
```diff
@@ -360,6 +361,14 @@ func (bc *BlockChain) CurrentFastBlock() *types.Block {
 	return bc.currentFastBlock.Load().(*types.Block)
 }
 
+// SetProcFeedback adds a feedback channel where true is sent each time block
```
Instead of the single channel, there should be a `event.Feed` for block processing start/end events. These events are useful in multiple places. For example, chain indexers could also be paused while block processing runs.

zsfelfoldi replied:
done

fjl commented on les/handler.go:
```diff
@@ -361,6 +463,43 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
+			if outSizeCost > cost {
+				cost = outSizeCost
+			}
+		}
```
Please put the 'real cost' computation into it's own function or method and document that it is the maximum of 'serving time', 'input size' and 'output size'.

fjl commented on les/server.go:
```diff
@@ -138,185 +258,145 @@ func (s *LesServer) SetBloomBitsIndexer(bloomIndexer *core.ChainIndexer) {
-	stats := make(map[uint64]*linReg)
-	for _, code := range reqList {
-		stats[code] = &linReg{cnt: 100}
+func (s *LesServer) gcfLoop() {
```
Suggestion: maybe put global cost factor calculation (and other cost stuff) into its own object like `costTracker`. Then you can even have cute names like `costTracker.globalFactor()` because it's obvious that all the code is about request cost calculation.

fjl commented on les/flowcontrol/logger.go:
```diff
@@ -0,0 +1,66 @@
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package flowcontrol implements a client side flow control mechanism
```
Please remove duplicated package doc comment.

fjl commented on les/api.go:
```diff
@@ -0,0 +1,176 @@
+// Note: priorityClientPool also stores a record about free clients while they are
+// connected in order to be able to assign priority to them later with the callback
+// function if necessary.
+func (v *priorityClientPool) connect(id enode.ID, updateCap func(uint64)) (uint64, bool) {
```
Right now the `priorityClientPool` relies on the `updateCap` callback to move the client to the free client pool. I think it would be better if the pools would coordinate clients among themselves. You need a reference to `freeClientPool` in `priorityClientPool` for that. Then `updateCap` would just be updating the capacity.

zsfelfoldi replied:
Done, I managed to removed this logic entirely from the handler. Now the pools implement the peerSetNotify interface (also have their own extended interface that also contains a setLimits call). I think it is much simpler and cleaner now.

zsfelfoldi commented on les/servingqueue.go:
```diff
@@ -0,0 +1,191 @@
+// Note: either blocking should be false or currentTask should be nil.
+func (sq *servingQueue) getNewTask(currentTask *servingTask, blocking bool) *servingTask {
+	sq.lock.Lock()
+	if sq.stopCount != 0 {
```
remove

zsfelfoldi commented:
@fjl PTAL
This is what I changed since the last review:
- implemented costTracker which now includes everything cost-related
- use fixed cost table with global correction factor
	- this also means that total capacity is now flexible which is a good thing so there is a little API extension to allow handling it nicely
	- api_test is also improved a bit to check the changing capacity and tolerate it
	- costTracker.globalFactor affects totalRecharge but there is also a capacity factor calculated in the client manager which is always >=1 and can allow more clients if the actual usage never exceeds the allowed totalRecharge amount
- factored out client pool logic from handler (free pool is now a child of the priority pool)
- use event.Feed for block processing start/stop events
- simplified the benchmark code a bit and added an API call
	- there was a kind of ugly logic for automatically creating an actual cost table based on measurements but it was dropped because now that is up to the user which will be us anyway (I think it is much cleaner to return measurement results directly and also allow any measurement setup than do some hard to follow black box magic and return the results only)
- wrote a simpler and cleaner serving queue with channels only (I think it is simple and elegant enough so it is better to have total control over such a sensitive part and be able to handle priorities as we wish)
	- the most important change is that the really ugly thing is gone where I needed to encapsulate the processing functions in sevingTask
	- now the processing can run in a simple goroutine and call three simple functions to start/optionally suspend/finish tasks (handler looks a lot better now, no convoluted way of thinking needed, also the database access-based suspend logic which you suggested can be realized later if we want it)

nambrot commented on les/handler.go:
```diff
@@ -286,17 +283,8 @@ func (pm *ProtocolManager) handle(p *peer) error {
 	}
-
-	if !pm.lightSync && !p.Peer.Info().Network.Trusted {
-		addr, ok := p.RemoteAddr().(*net.TCPAddr)
```
So this piece of code existed to IIUC to assign some kind of reputation to IP addresses, because otherwise light clients could just create a new ID and further free-ride. I'm curious what led to the removal of this code, and if we have coverage for malbehavior instead?
